docs/source/fine-performance-metrics.rst

```diff
@@ -0,0 +1,198 @@
+<https://github.com/dask/dask-labextension>`_, you can just drag the
+``Fine Performance Metrics`` widget onto your Jupyter dashboard.
+
+.. image:: images/fine-performance-metrics/seconds-full.png
+   :alt: Populated Fine Performance Metrics dashboard
+
+The central panel shows, of all the time that the cluster spent doing something, what
+*activities* it spent performing. The most important ones are:
+
+``thread-cpu``
+  CPU time spent by tasks while running on workers
+``thread-noncpu``
+  Difference between wall clock time and CPU time spent by tasks while running on
+  workers. This is typically either I/O time, CPU contention, or GIL contention.
+``idle``
+  Time where a worker had a free thread, but nothing to run on it. This is typically
+  caused by the workload not being able to fully utilize all threads on the cluster,
+  excessive CPU load on the scheduler, or network latency between scheduler and workers.
+  This measure does not include time spent while the whole cluster was completely idle.
+
+The grand total of the time shown should roughly add up to the end-to-end runtime of
+your workload, multiplied by the number of threads on the cluster.
+
+The plot on the right shows network transfer time. Note that most of it should be
+pipelined with task execution, so it may not have an impact. You should worry about this
+only if you have a very large ``idle`` time.
+
+The plot on the left shows the same information as the central one, but broken down by
+function. There is a filter that allows you to show only selected functions. In the
+sample screenshots, you can observe that most of the `thread-noncpu` time is
+concentrated - as expected - in functions that are known to be I/O heavy. Here they are
+singled out:
+
+.. image:: images/fine-performance-metrics/seconds-IO.png
+   :alt: Fine Performance Metrics dashboard, just the I/O functions
+
+And here's everything but those three I/O heavy functions:
+
+.. image:: images/fine-performance-metrics/seconds-not-IO.png
+   :alt: Fine Performance Metrics dashboard, except the I/O functions
+
+This tells us an important piece of information: why is `astype`, which is a pure CPU
+function, spending so much time occupying workers' threads but without accruing any CPU
+time? The answer, almost certainly, it's that it doesn't properly release the GIL.
+
+Fine Performance Metrics don't collect exclusively wall clock timings. We can change the
+unit to bytes:
+
+.. image:: images/fine-performance-metrics/bytes.png
+   :alt: Fine Performance Metrics dashboard, unit=bytes
+
+The above gives us insights on spill/unspill activity (see :doc:`worker-memory`): in
+this workflow, in 99% of the cases there was enough RAM to accommodate all the data and
+it was unnecessary to retrieve it from disk; in other words we have a 99% *cache hit
+ratio*, which means that we would not get any benefit if we increased RAM, but we would
+likely start seeing slowdowns if we reduced it.
```

Comment:
This is conflating cache hits and byte hits. Since we're counting bytes here, this is less about the number of times the cache got hit but rather about the amount of data we were able to retrieve from the cache. Maybe this is clearer?

```suggestion
this workflow, 99% of the data was found in memory and it did not have to be retrieved from disk; in other words, we have a 99% *byte hit
ratio*, which means that we would not get much benefit if we increased RAM, but we would
likely start seeing slowdowns if we reduced it.
```

Reply:
Yes, "bytes" shows cache hit ratio weighted by data size, whereas "count" shows unweighted cache hit ratio e.g. number of accesses. Unless your chunk size is extremely small and you are backed by a disk with very high latency (e.g. an NFS mount), you typically only care about the former.

This said, `cache hit ratio` immediately evokes the correct image in the mind of any reader, whereas `byte hit ratio` definitely doesn't.

Reply:
> This said, `cache hit ratio` immediately evokes the correct image in the mind of any reader, whereas `byte hit ratio` definitely doesn't.

I'd disagree that in the current form it evokes the correct image when thinking about uneven data sizes. For example, if the 1 % of data stem from a single but huge result while we have _many_ small ones, it currently evokes the image that we had to load from cache 1 out of 100 intermediate results. Instead, we had to read only one out of thousands or more. If we want to avoid `byte hit ratio` as a fairly uncommon term, I'd suggest stressing the point that it's weighted by data size. Maybe this is as simple as 
```suggestion
this workflow, only 1% of the data had to be retrieved from disk, whereas 99% of the data could be stored in memory; in other words we have a *cache hit* for 99% of the data, which means that we would not get any benefit if we increased RAM, but we would
likely start seeing slowdowns if we reduced it.
```

Reply:
Either way, this is non-blocking.

Reply:
tweaked

Reply:
Thanks!